src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -241,19 +241,275 @@ uct_ib_mlx5_devx_reg_ksm_data(uct_ib_mlx5_md_t *md,
+    kh_init_inplace(rkeys, &md->lru_rkeys.hash);
+    md->lru_rkeys.counter = 0;
+}
+
+static void uct_ib_mlx5_devx_mr_lru_cleanup(uct_ib_mlx5_md_t *md)
+{
+    while (kh_size(&md->lru_rkeys.hash) > 0) {
+        uct_ib_mlx5_devx_md_mr_lru_pop(md);
+    }
+
+    kh_destroy_inplace(rkeys, &md->lru_rkeys.hash);
+}
+
+static ucs_status_t uct_ib_mlx5_devx_reg_indirect_key(uct_ib_md_t *ibmd,
+                                                      uct_ib_mem_t *ib_memh)
+{
+    uct_ib_mlx5_md_t *md    = ucs_derived_of(ibmd, uct_ib_mlx5_md_t);
+    uct_ib_mlx5_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
+    uint8_t *is_dummy       = NULL;
+    unsigned i;
+    ucs_status_t status;
+
+    if (!(md->flags & UCT_IB_MLX5_MD_FLAG_KSM)) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    i = 0;
+    do {
+        status = uct_ib_mlx5_devx_reg_ksm_data_contig(md,
+                                                      &memh->mrs[UCT_IB_MR_DEFAULT],
+                                                      0, &memh->indirect_dvmr,
+                                                      &memh->super.indirect_rkey);
+        if (ucs_unlikely(status != UCS_OK)) {
+            break;
+        }
+
+        status = uct_ib_md_mlx5_devx_mr_lru_push(md, memh->super.indirect_rkey,
+                                                 memh->indirect_dvmr, &is_dummy);
+        ++i;
+    } while ((i < ibmd->config.max_mr_gc_queue) &&
+             (status == UCS_ERR_ALREADY_EXISTS));
+
+    if (ucs_unlikely(status != UCS_OK)) {
+        goto err;
+    }
+
+    ucs_assert(is_dummy != NULL);
+    *is_dummy = 0;
+    return UCS_OK;
+err:
+
+    ucs_assert(memh->super.indirect_rkey == UCT_IB_INVALID_MKEY);
+    if (status == UCS_ERR_UNSUPPORTED) {
+        md->flags &= ~UCT_IB_MLX5_MD_FLAG_KSM;
+    } else if(status == UCS_ERR_ALREADY_EXISTS) {
+        ucs_error("%s: rkeys LRU cache size=%d is too small",
+                  uct_ib_device_name(&ibmd->dev),
+                  ibmd->config.max_mr_gc_queue);
+        status = UCS_ERR_NO_MEMORY;
+    }
+
+    return status;
+}
+
+static ucs_status_t uct_ib_mlx5_devx_dereg_key(uct_ib_md_t *ibmd,
+                                               uct_ib_mem_t *ib_memh,
+                                               uct_ib_mr_type_t mr_type)
+{
+    uct_ib_mlx5_md_t *md    = ucs_derived_of(ibmd, uct_ib_mlx5_md_t);
+    uct_ib_mlx5_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
+    ucs_status_t ret_status = UCS_OK;
+    uint8_t *is_dummy       = NULL;
+    ucs_status_t status;
+    int ret;
+
+    if ((md->flags & UCT_IB_MLX5_MD_FLAG_KSM) &&
+        (memh->super.indirect_rkey != UCT_IB_INVALID_MKEY)) {
+        status = uct_ib_md_mlx5_devx_mr_lru_push(md, memh->super.indirect_rkey,
+                                                 NULL, &is_dummy);
+        if (((status == UCS_OK) ||
+             (status == UCS_ERR_ALREADY_EXISTS)) && is_dummy) {
+            *is_dummy = 1;
+        }
```

Comment:
probably uct_ib_md_mlx5_devx_mr_lru_push should set dummy=1 whenever with push mr=NULL for existing key